tests/docker/docker_test.go

```diff
@@ -0,0 +1,219 @@
+	tc           *projects.TestContext
+	externalIP   string
+	appName      string
+	externalPort = 8028
+	appLink      = "docker://nginx"
+	portPublish  = []string{fmt.Sprintf("%d:80", externalPort)}
+)
+
+// TestMain is used to provide setup and teardown for the rest of the
+// tests. As part of setup we make sure that context has a slice of
+// EVE instances that we can operate on. For any action, if the instance
+// is not specified explicitly it is assumed to be the first one in the slice
+func TestMain(m *testing.M) {
+	fmt.Println("Docker app deployment Test")
+
+	tc = projects.NewTestContext()
+
+	projectName := fmt.Sprintf("%s_%s", "TestLogInfoMetric", time.Now())
+
+	tc.InitProject(projectName)
+
+	for _, node := range tc.GetNodeDescriptions() {
+		edgeNode := tc.GetController().GetEdgeNode(node.Name)
+		if edgeNode == nil {
+			edgeNode = tc.NewEdgeNode(tc.WithNodeDescription(node), tc.WithCurrentProject())
+		} else {
+			tc.UpdateEdgeNode(edgeNode, tc.WithCurrentProject(), tc.WithDeviceModel(node.Model))
+		}
+
+		tc.ConfigSync(edgeNode)
+
+		if edgeNode.GetState() == device.NotOnboarded {
+			log.Fatal("Node is not onboarded now")
+		}
+
+		tc.AddNode(edgeNode)
+	}
+
+	tc.StartTrackingState(false)
+
+	res := m.Run()
+
+	os.Exit(res)
+}
+
+//checkAppMsg wait for info of ZInfoApp type with mention of deployed AppName
+func checkAppMsg(appName string) projects.ProcInfoFunc {
+	return func(msg *info.ZInfoMsg) error {
+		if msg.Ztype == info.ZInfoTypes_ZiApp {
+			if msg.GetAinfo().AppName == appName {
+				return fmt.Errorf("app found with name %s", appName)
+			}
+		}
+		return nil
+	}
+}
+
+//checkAppRunning wait for info of ZInfoApp type with mention of deployed AppName and ZSwState_RUNNING state
+func checkAppRunning(appName string) projects.ProcInfoFunc {
+	return func(msg *info.ZInfoMsg) error {
+		if msg.Ztype == info.ZInfoTypes_ZiApp {
+			if msg.GetAinfo().AppName == appName {
+				if msg.GetAinfo().State == info.ZSwState_RUNNING {
+					return fmt.Errorf("app RUNNING with name %s", appName)
+				}
+			}
+		}
+		return nil
+	}
+}
+
+//requestApp try to get info from http endpoint of APP
+func requestApp(ip string) error {
+	res, err := utils.RequestHTTPWithTimeout(fmt.Sprintf("http://%s:%d", ip, externalPort), time.Second)
+	if err != nil {
+		return nil
+	} else {
+		return fmt.Errorf(res)
+	}
+}
+
+//requestApp try to get info from http endpoint of APP
+func getEVEIP(edgeNode *device.Ctx) projects.ProcTimerFunc {
+	return func() error {
+		if edgeNode.GetDevModel() == defaults.DefaultRPIModel {
+			eveIPs := tc.GetState(edgeNode).GetEVEIPs()
+			if len(eveIPs) > 0 {
+				externalIP = eveIPs[0]
+				return fmt.Errorf("external ip is: %s", externalIP)
+			}
+		} else {
+			externalIP = "127.0.0.1"
+			return fmt.Errorf("external ip is: %s", externalIP)
+		}
+		return nil
+	}
+}
+
+//checkAppAccess try to access APP with timer
+func checkAppAccess() projects.ProcTimerFunc {
```

Comment:
It looks complicated... 2 functions (this & requestApp ) to check app availabily? Can we make it 1?

Reply:
functions merged